File: apps/web/src/services/weather.ts
import { InsightType, PaginatedResponseType, WeatherLogType } from '@repo/shared'

import { api } from '@/lib/api'

export interface CurrentWeather {
  temperature: number
  humidity: number
  windSpeed: number
  condition: string
  rainProbability: number
  location: string
}

const weatherConditions: Record<number, string> = {
  0: 'Céu limpo',
  1: 'Principalmente limpo',
  2: 'Parcialmente nublado',
  3: 'Nublado',
  45: 'Névoa',
  48: 'Névoa com geada',
  51: 'Garoa leve',
  53: 'Garoa moderada',
  55: 'Garoa intensa',
  61: 'Chuva leve',
  63: 'Chuva moderada',
  65: 'Chuva forte',
  71: 'Neve leve',
  73: 'Neve moderada',
  75: 'Neve forte',
  80: 'Pancadas leves',
  81: 'Pancadas moderadas',
  82: 'Pancadas fortes',
  95: 'Tempestade',
  96: 'Tempestade com granizo',
  99: 'Tempestade forte com granizo',
}

export async function fetchWeatherByCoordinates(
  latitude: number,
  longitude: number,
  locationName: string
): Promise<CurrentWeather> {
  const url = 'https://api.open-meteo.com/v1/forecast'
  const params = new URLSearchParams({
    latitude: latitude.toString(),
    longitude: longitude.toString(),
    current: 'temperature_2m,relative_humidity_2m,weather_code,wind_speed_10m,precipitation_probability',
    timezone: 'America/Sao_Paulo',
  })

  const response = await fetch(`${url}?${params}`)
  const data = await response.json()

  const current = data.current || {}

  return {
    temperature: current.temperature_2m || 0,
    humidity: current.relative_humidity_2m || 0,
    windSpeed: current.wind_speed_10m || 0,
    condition: weatherConditions[current.weather_code] || 'Desconhecido',
    rainProbability: current.precipitation_probability || 0,
    location: locationName,
  }
}

export async function fetchWeatherLogs(
  page = 1,
  limit = 10,
  location?: string,
): Promise<PaginatedResponseType<WeatherLogType>> {
  const response = await api.get('/api/weather/logs', { params: { page, limit, location } })
  return response.data
}

export async function fetchLatestWeather(location?: string): Promise<WeatherLogType | null> {
  const response = await api.get('/api/weather/logs/latest', { params: { location } })
  return response.data
}

export async function collectWeather(
  latitude: number,
  longitude: number,
  location: string,
): Promise<WeatherLogType> {
  const response = await api.post('/api/weather/collect', { latitude, longitude, location })
  return response.data
}

export async function fetchInsights(): Promise<InsightType[]> {
  const response = await api.get('/api/weather/insights')
  return response.data
}

export async function generateInsights(): Promise<InsightType[]> {
  const response = await api.post('/api/weather/insights/generate')
  return response.data
}

export function getExportCsvUrl(): string {
  return `${import.meta.env.VITE_API_BASE_URL}/api/weather/export/csv`
}

export function getExportXlsxUrl(): string {
  return `${import.meta.env.VITE_API_BASE_URL}/api/weather/export/xlsx`
}

File: services/python-service/src/weather.py
import requests
import logging
from geopy.geocoders import Nominatim
from datetime import datetime
from pytz import timezone

# Configurar logging
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s'
)
logger = logging.getLogger(__name__)


WEATHER_CODE = {
    0: "Céu limpo",
    1: "Principalmente limpo",
    2: "Parcialmente nublado",
    3: "Nublado",
    45: "Nevoeiro",
    48: "Nevoeiro com gelo",
    51: "Garoa leve",
    53: "Garoa",
    55: "Garoa intensa",
    61: "Chuva leve",
    63: "Chuva moderada",
    65: "Chuva forte",
    80: "Aguaceiros fracos",
    81: "Aguaceiros moderados",
    82: "Aguaceiros fortes",
} # Código de condição climática baseado na API Open-Meteo


def get_coordinates():
    geolocator = Nominatim(user_agent="coordenadas_teresina")
    cidade = "Teresina, Piauí, Brasil"
    try:
        location = geolocator.geocode(cidade, timeout=20)
        if location is None:
            raise ValueError("Não foi possível obter as coordenadas para a cidade especificada.")
        
        cordinates = {
            "latitude": location.latitude,
            "longitude": location.longitude
        }
        logger.info(f"Coordenadas obtidas com sucesso para {cidade}")
        return cordinates
    except ValueError as e:
        logger.error(f"Erro ao obter coordenadas: {e}")
        return None
    except Exception as e:
        logger.error(f"Erro inesperado ao obter coordenadas: {e}")
        return None


def search_forecast(lat: float, lon: float, tz: str):
    url = "https://api.open-meteo.com/v1/forecast"
    params = {
        "latitude": lat,
        "longitude": lon,
        "timezone": tz,
        "current_weather": True,
        "hourly": (
            "temperature_2m,"
            "relative_humidity_2m,"
            "precipitation_probability"
        )
    }
    try:
        response = requests.get(url, params=params, timeout=10)
        response.raise_for_status()
        logger.info("Previsão do tempo obtida com sucesso da API Open-Meteo")
        return response.json()
    except requests.exceptions.RequestException as e:
        logger.error(f"Erro ao buscar previsão do tempo: {e}")
        raise


def get_time_info():
    try:
        cordinates = get_coordinates()
        if not cordinates:
            logger.error("Não foi possível obter coordenadas")
            return None
        
        forecast = search_forecast(cordinates["latitude"], cordinates["longitude"], "America/Fortaleza")
        data = extract_data(forecast)
        formatted_data = format_data(**data)
        logger.info("Dados meteorológicos processados com sucesso")
        return formatted_data
    
    except Exception as e:
        logger.error(f"Erro ao obter informações meteorológicas: {e}", exc_info=True)
        return None


def extract_data(func):
    current = func["current_weather"]
    hourly = func["hourly"]
    weather_code = current["weathercode"]
    condition = WEATHER_CODE.get(weather_code, "Desconhecido")


    current_time = current["time"]
    times = hourly["time"]
    idx = 0
    for i, t in enumerate(times):
        if t <= current_time:
            idx = i
        else:
            break

    return {
        "current": current,
        "hourly": hourly,
        "condition": condition,
        "precipitation_idx": idx
    }


def format_data(**kwargs):
    tz_brasil = timezone("America/Fortaleza")
    data_coleta = datetime.now(tz=tz_brasil).isoformat()
    idx = kwargs.get("precipitation_idx", 0)
    data_formatted = {
        "temperatura": kwargs["current"]["temperature"],
        "umidade": float(kwargs['hourly']['relative_humidity_2m'][idx]),
        "vento": kwargs["current"]["windspeed"],
        "condicao": kwargs["condition"],
        "probabilidade_chuva": float(kwargs['hourly']['precipitation_probability'][idx]),
        "data_coleta": data_coleta,
        "cidade": "Teresina"
    }

    return data_formatted